src/samplers/hmc.jl

```diff
@@ -169,33 +172,43 @@ end
 
 function assume(spl :: HMCSampler{HMC}, d :: Distribution, prior :: Prior)
   dprintln(2, "assuming...")
+  # 1. Gen priors and vectorize if necessary
+
   # TODO: Change the first running condition
   # If it's the first time running the program
   if spl.first
+    dprintln(2, "generating priors...")
     # Generate a new prior
     r = rand(d)
-    dim = length(r)
-    if dim == 1
+    if prior.typ == 1
       val = Vector{Any}([Dual(r)])
-    else
+    elseif prior.typ == 2
       val = Vector{Any}(map(x -> Dual(x), r))
+    elseif prior.typ == 3
+      val = Vector{Any}(map(x -> Dual(x), vec(r)))
     end
     # Store the generated prior
     addPrior(spl.priors, prior, val)
   # If not the first time
   else
     # Fetch the existing prior
+    dprintln(2, "fetching priors...")
     val = spl.priors[prior]
   end
-  if isa(val, Array)
-    if length(val) == 1
-      # Turn Array{Any} to Any if necessary (this is due to randn())
-      val = val[1]
-    else
-      # Turn Array{Any} to Array{T} if necessary (this is due to an update in Distributions.jl)
-      T = typeof(val[1])
-      val = Vector{T}(val)
-    end
+
+  # 2. reconstruct priors
+  dprintln(2, "reconstructing priors...")
+  if prior.typ == 1
+    # Turn Array{Any} to Any if necessary (this is due to randn())
+    val = val[1]
+  elseif prior.typ == 2
+    # Turn Vector{Any} to Vector{T} if necessary (this is due to an update in Distributions.jl)
+    T = typeof(val[1])
+    val = Vector{T}(val)
+  elseif prior.typ == 3
+    T = typeof(val[1])
+    dim = Int(sqrt(length(val)))
```

Comment:
Not all matrix-variates are square matrices, we might want to store the `dim` information in `Prior` as well.